File: templates/problems/06_linked_list/01_reverse_list/typescript/reverse_list.test.ts
import { describe, it, expect } from "vitest";
import { buildList, listToSlice } from "./helpers";
import { reverseList } from "./reverse_list";

describe("reverseList", () => {
    it("nil list", () => {
        expect(listToSlice(reverseList(buildList([])))).toEqual([]);
    });

    it("single element", () => {
        expect(listToSlice(reverseList(buildList([1])))).toEqual([1]);
    });

    it("two elements", () => {
        expect(listToSlice(reverseList(buildList([1, 2])))).toEqual([2, 1]);
    });

    it("multiple elements", () => {
        expect(listToSlice(reverseList(buildList([1, 2, 3, 4, 5])))).toEqual([5, 4, 3, 2, 1]);
    });

    it("duplicates", () => {
        expect(listToSlice(reverseList(buildList([1, 1, 2, 2])))).toEqual([2, 2, 1, 1]);
    });

    it("negative values", () => {
        expect(listToSlice(reverseList(buildList([-1, 0, 1])))).toEqual([1, 0, -1]);
    });
});

File: templates/problems/06_linked_list/01_reverse_list/go/reverse_list_test.go
package reverse_list

import (
	"reflect"
	"testing"
)

func TestReverseList(t *testing.T) {
	tests := []struct {
		name string
		vals []int
		want []int
	}{
		{"nil list", nil, nil},
		{"single element", []int{1}, []int{1}},
		{"two elements", []int{1, 2}, []int{2, 1}},
		{"multiple elements", []int{1, 2, 3, 4, 5}, []int{5, 4, 3, 2, 1}},
		{"duplicates", []int{1, 1, 2, 2}, []int{2, 2, 1, 1}},
		{"negative values", []int{-1, 0, 1}, []int{1, 0, -1}},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			head := buildList(tt.vals)
			got := listToSlice(reverseList(head))
			if !reflect.DeepEqual(got, tt.want) {
				t.Errorf("reverseList() = %v, want %v", got, tt.want)
			}
		})
	}
}
